### PR TITLE
treewide: Initial role PR

### PR DESCRIPTION
**Pull request by haslersn**
* [x] Don't hardcode the python version
* [x] Add `README.md`

### CODE REVIEW

File: defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      format: '%(asctime)s - %(name)s - %(lineno)d - %(levelname)s - %(request)s - %(message)s'
+  filters:
+    context:
+      (): synapse.util.logcontext.LoggingContextFilter
+      request: ""
+  handlers:
+    file:
+      class: logging.handlers.RotatingFileHandler
+      formatter: precise
+      filename: "{{ synapse_data_dir }}/homeserver.log"
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]: Please place logs in `/var/log/synapse`, not in `/etc` or user-configurable
```suggestion
      filename: "/var/log/synapse/homeserver.log"
```

File: defaults/main.yml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      maxBytes: 104857600
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]:
```suggestion
      maxBytes: 104857600 # 100MiB
```

File: defaults/main.yml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      backupCount: 10
+      filters: [context]
+      encoding: utf8
+    console:
```
Comment (SlothOfAnarchy) [inline defaults/main.yml:24]: Will this log to systemd's journal?

File: defaults/main.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  loggers:
+    synapse:
+      level: INFO
+    synapse.storage.SQL:
+      # beware: increasing this to DEBUG will make synapse log sensitive
+      # information such as access tokens.
+      level: INFO
+  root:
+    level: INFO
+    handlers: [file, console]
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]:
```suggestion
    handlers:
      - file
      - console
```

File: defaults/main.yml [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      (): synapse.util.logcontext.LoggingContextFilter
+      request: ""
+  handlers:
+    file:
+      class: logging.handlers.RotatingFileHandler
+      formatter: precise
+      filename: "{{ synapse_data_dir }}/homeserver.log"
+      maxBytes: 104857600
+      backupCount: 10
+      filters: [context]
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]:
```suggestion
      filters:
        - context
```

File: defaults/main.yml [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      encoding: utf8
+    console:
+      class: logging.StreamHandler
+      formatter: precise
+      filters: [context]
```
Comment (SlothOfAnarchy) [inline defaults/main.yml]:
```suggestion
      filters:
        - context
```

File: tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    name: "{{ synapse_user }}"
+    state: present
+
+- name: Prepare Synapse directories
+  file:
+    path: "{{ item }}"
+    state: directory
+    owner: "{{ synapse_user }}"
+  loop:
+    - "{{ synapse_data_dir }}"
```
Comment (SlothOfAnarchy) [inline tasks/main.yml:14]:
```suggestion
    - "{{ synapse_data_dir }}"
    - /var/log/synapse
```

File: tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      - pip
+      - setuptools
+      - matrix-synapse
+  notify:
+    - Restart Synapse service
+
+- name: Install Synapse REST Password provider
+  get_url:
+    url: https://raw.githubusercontent.com/kamax-matrix/matrix-synapse-rest-auth/master/rest_auth_provider.py
+    dest: "{{ synapse_install_dir }}/env/lib/python3.5/site-packages/rest_auth_provider.py"
```
Comment (SlothOfAnarchy) [inline tasks/main.yml]: Please don't hardcode the python version

**Review by SlothOfAnarchy — DISMISSED** [review]

File: tasks/main.yml (hunk 8 above)
Comment (haslersn) [reply tasks/main.yml]: We could read the site-packages from the virtualenv via `{{ synapse_install_dir }}/env/bin/python -m site`, but the output still has to be parsed.

But how does this work together with upgrading the python version? As it turns out, the ansible pip module (with virtualenv parameter) will not upgrade the python version of an existing virtualenv. For a freshly created virtualenv, by the way, it uses (by default) the python version which ansible has been run with (on the remote). This is in our setup the (on the remote) system-wide installed python3 version.

Of course I think that whether or not the virtualenv is freshly created shouldn't make a difference, for the sake of purity. Therefore we first need to find a solution on how to upgrade the virtualenv, as soon as a new python3 version is deployed system-wide. This solution should of course **not** encompass deleting and recreating the virtualenv on every deployment.

I think that we should solve that problem first. Its solution might change the possibilities how the problem with the hardcoded python version can be solved.

Comment (haslersn) [reply tasks/main.yml]: Or we could just omit the virtualenv...

Comment (haslersn) [reply tasks/main.yml]: This is now resolved. See my newest commit on this branch: 1a59cf84dcfaf84935d23f9d8bdc8106731373f0

File: defaults/main.yml (hunk 3 above)
Comment (haslersn) [reply defaults/main.yml:24]: Yes

File: tasks/main.yml (hunk 8 above)
Comment (haslersn) [reply tasks/main.yml]: I cannot guarantee that the downloaded `matrix-synapse-rest-auth` plugin will work with other Python versions, though. Do you approve that this is resolved, nonetheless?

**Review by Mr-Pi — APPROVED** [review]

File: README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+## Role variables
+
+| Name                        | Default                 | Description                                    |
+| :-------------------------- | :---------------------- | :--------------------------------------------- |
+| `synapse_data_dir`          | `/etc/synapse`          | Directory to store config and database         |
+| `synapse_homeserver_config` |                         | [#configuration](#configuration) |
+| `synapse_install_dir`       | `/opt/synapse`          | Directory to store the synapse installation    |
+| `synapse_log_config`        | See `defaults/main.yml` | Content of `<server_name>.log.config` |
+| `synapse_signing_key`       |                         | [#configuration](#configuration) |
```
Comment (SlothOfAnarchy) [inline README.md:50]: Please indent

File: defaults/main.yml [hunk 10]
```diff
@@ -0,0 +1,40 @@
+---
+synapse_user: synapse
+synapse_install_dir: /opt/synapse
+synapse_data_dir: /etc/synapse
```
Comment (SlothOfAnarchy) [inline defaults/main.yml:4]: `/etc` is for configuration files, see https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard#Directory_structure

File: templates/synapse.service.j2 [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# {{ ansible_managed }}
+[Unit]
+Description=Matrix Synapse homeserver
+
+[Service]
+Type=notify
+NotifyAccess=main
+User={{ synapse_user }}
+ExecStart={{ synapse_fact_virtualenv_dir }}/bin/python -m synapse.app.homeserver --config-path={{ synapse_data_dir }}/homeserver.yaml
```
Comment (SlothOfAnarchy) [inline templates/synapse.service.j2:9]: not `python3`?

Comment (haslersn) [reply templates/synapse.service.j2:9]: No, there's only one python version in a virtualenv.

**Pull request merged**